k9mail/src/main/java/com/fsck/k9/activity/UnreadWidgetConfiguration.java

```diff
@@ -41,58 +61,153 @@ public void onCreate(Bundle icicle) {
+            @Override
+            public boolean onPreferenceClick(Preference preference) {
+                Intent intent = new Intent(UnreadWidgetConfiguration.this, ChooseFolder.class);
+                intent.putExtra(ChooseFolder.EXTRA_ACCOUNT, selectedAccountUuid);
+                intent.putExtra(ChooseFolder.EXTRA_SHOW_DISPLAYABLE_ONLY, "yes");
+                startActivityForResult(intent, REQUEST_CHOOSE_FOLDER);
+                return false;
+            }
+        });
         setTitle(R.string.unread_widget_select_account);
     }
 
     @Override
-    protected boolean displaySpecialAccounts() {
+    public void onActivityResult(int requestCode, int resultCode, Intent data) {
+        if (resultCode == RESULT_OK) {
+            switch (requestCode) {
+                case REQUEST_CHOOSE_ACCOUNT:
+                    selectedAccountUuid = data.getStringExtra(ChooseAccount.EXTRA_ACCOUNT_UUID);
+                    unreadFolder.setSummary(getString(R.string.unread_widget_folder_summary));
+                    if (SearchAccount.UNIFIED_INBOX.equals(selectedAccountUuid) ||
+                            SearchAccount.ALL_MESSAGES.equals(selectedAccountUuid)) {
+                        unreadAccount.setSummary(selectedAccountUuid);
+                        unreadFolderEnabled.setEnabled(false);
+                        unreadFolder.setEnabled(false);
+                        selectedFolderName = null;
+                    } else {
+                        Account selectedAccount = Preferences.getPreferences(this).getAccount(selectedAccountUuid);
+                        unreadAccount.setSummary(selectedAccount.getName());
+                        unreadFolderEnabled.setEnabled(true);
+                        unreadFolder.setEnabled(true);
+                    }
+                    break;
+                case REQUEST_CHOOSE_FOLDER:
+                    selectedFolderName = data.getStringExtra(ChooseFolder.EXTRA_NEW_FOLDER);
+                    unreadFolder.setSummary(selectedFolderName);
+                    break;
+            }
+        }
+        super.onActivityResult(requestCode, resultCode, data);
+    }
+
+    @Override
+    public boolean onCreateOptionsMenu(Menu menu) {
+        super.onCreateOptionsMenu(menu);
+        getMenuInflater().inflate(R.menu.unread_widget_option, menu);
         return true;
     }
 
     @Override
-    protected void onAccountSelected(BaseAccount account) {
-        // Save widget configuration
-        String accountUuid = account.getUuid();
-        saveAccountUuid(this, mAppWidgetId, accountUuid);
+    public boolean onOptionsItemSelected(MenuItem item) {
+        switch (item.getItemId()) {
+            case R.id.done:
+                checkToCreateWidget();
+                return true;
+            default:
+                return super.onOptionsItemSelected(item);
+        }
+    }
+
+    private void checkToCreateWidget() {
```

Comment:
`validateThenUpdateWidgetThenExitIfPossible()` would be more accurate - you're doing more than just checking here.

Alternatively, make this method return a value and then do


```
if(validateWidget()) {
   updateWidgetAndExit();
}
```

in `onOptionsItemSelected`